pkg/cloudprovider/aws/eventbridge.go

```diff
@@ -0,0 +1,190 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package aws
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"sync"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/eventbridge"
+	"github.com/aws/aws-sdk-go/service/eventbridge/eventbridgeiface"
+	"github.com/samber/lo"
+	"go.uber.org/multierr"
+
+	"github.com/aws/karpenter/pkg/apis/provisioning/v1alpha5"
+	"github.com/aws/karpenter/pkg/utils/injection"
+)
+
+type EventBridgeProvider struct {
+	client    eventbridgeiface.EventBridgeAPI
+	queueName string
+	metadata  *Metadata
+}
+
+type EventRule struct {
+	Name    string
+	Pattern *EventPattern
+	Target  *EventTarget
+}
+
+type EventTarget struct {
+	ID  string
+	ARN string
+}
+
+type EventPattern struct {
+	Source     []string `json:"source,omitempty"`
+	DetailType []string `json:"detail-type,omitempty"`
+}
+
+func (ep *EventPattern) Serialize() []byte {
+	return lo.Must(json.Marshal(ep))
+}
+
+func NewEventBridgeProvider(eb eventbridgeiface.EventBridgeAPI, metadata *Metadata, queueName string) *EventBridgeProvider {
+	return &EventBridgeProvider{
+		client:    eb,
+		metadata:  metadata,
+		queueName: queueName,
+	}
+}
+
+func (eb *EventBridgeProvider) CreateEC2NotificationRules(ctx context.Context) (err error) {
+	wg := &sync.WaitGroup{}
```

Comment:
I thought the same after a bit, actually implemented an `atomic.MultiErr` but then realized that this creates weird behavior when it comes to checking the "nullity" of errors based on convention.

```go
err := &atomic.MultiErr{}
go func(){
  e := DoSomeAction()
  err.Append(e)
}
go func(){
  e := DoSomeOtherAction()
  err.Append(e)
}
// This is the incorrect call but feels natural based on how errs are typically checked
if err != nil {
  // Comes here even if there is no error
}
```

I assume a better abstraction is to just abstract the entire goroutine process similary to `golang.org/x/sync/errgroup` but I wanted to return the full error set rather than just the first error. Probably just roll our own abstraction. It's not that difficult.

Reply:
Using `workqueue.ParallelizeUntil` instead